code/modules/mob/living/carbon/xenomorph/castes/baneling/abilities_baneling.dm

```diff
@@ -0,0 +1,127 @@
+// ***************************************
+// *********** Baneling Explode
+// ***************************************
+/datum/action/xeno_action/baneling_explode
+	name = "Baneling Explode"
+	action_icon_state = "baneling_explode"
+	desc = ""
+	ability_name = "baneling explode"
+	keybinding_signals = list(
+		KEYBINDING_NORMAL = COMSIG_XENOABILITY_BANELING_EXPLODE,
+	)
+
+/datum/action/xeno_action/baneling_explode/give_action(mob/living/L)
+	. = ..()
+	var/mob/living/carbon/xenomorph/baneling/X = L
+	RegisterSignal(X, COMSIG_MOB_DEATH, PROC_REF(handle_smoke))
+
+/datum/action/xeno_action/baneling_explode/action_activate()
+	. = ..()
+	var/mob/living/carbon/xenomorph/baneling/X = owner
+	handle_smoke(ability = TRUE)
+	X.death(FALSE)
+
+/datum/action/xeno_action/baneling_explode/proc/handle_smoke(smoke_duration, datum/effect_system/smoke_spread/smoke, smoke_range, ability = FALSE)
```

Comment:
Needs to be a SIGNAL_HANDLER since you're calling it via a signal register. None of the first three arguments are used now that it doesn't loop.